llvm/include/llvm/Transforms/Utils/SYCLModuleSplit.h

```diff
@@ -0,0 +1,224 @@
+  std::unique_ptr<Module> releaseModulePtr() { return std::move(M); }
+
+  // Sometimes, during module transformations, some Function objects within the
+  // module are replaced with different Function objects with the same name.
+  // Entry points need to be updated to include the replacement function.
+  // save/rebuild pair of functions is provided to automate this process.
+  void saveEntryPointNames(std::vector<std::string> &Dest) {
+    EntryPoints.saveNames(Dest);
+  }
+
+  void rebuildEntryPoints(const std::vector<std::string> &Names) {
+    EntryPoints.rebuildFromNames(Names, getModule());
+  }
+
+  void rebuildEntryPoints(const Module &M) { EntryPoints.rebuild(M); }
+
+  void rebuildEntryPoints() { EntryPoints.rebuild(*M); }
+
+  // Cleans up module IR - removes dead globals, debug info etc.
+  void cleanup();
+
+  ModuleDesc clone() const;
+
+  std::string makeSymbolTable() const;
+
+  void dump(raw_ostream &OS) const;
+};
+
+// Module split support interface.
+// It gets a module (in a form of module descriptor, to get additional info) and
+// a collection of entry points groups. Each group specifies subset entry points
+// from input module that should be included in a split module.
+class ModuleSplitterBase {
+protected:
+  ModuleDesc Input;
+  EntryPointGroupVec Groups;
+
+protected:
+  EntryPointGroup nextGroup() {
+    assert(hasMoreSplits() && "Reached end of entry point groups list.");
+    EntryPointGroup Res = std::move(Groups.back());
+    Groups.pop_back();
+    return Res;
+  }
+
+  Module &getInputModule() { return Input.getModule(); }
+
+  std::unique_ptr<Module> releaseInputModule() {
+    return Input.releaseModulePtr();
+  }
+
+public:
+  ModuleSplitterBase(ModuleDesc MD, EntryPointGroupVec GroupVec)
+      : Input(std::move(MD)), Groups(std::move(GroupVec)) {
+    assert(!Groups.empty() && "Entry points groups collection is empty!");
+  }
+
+  virtual ~ModuleSplitterBase() = default;
+
+  // Gets next subsequence of entry points in an input module and provides split
+  // submodule containing these entry points and their dependencies.
+  virtual ModuleDesc nextSplit() = 0;
+
+  // Returns a number of remaining modules, which can be split out using this
+  // splitter. The value is reduced by 1 each time nextSplit is called.
+  size_t remainingSplits() const { return Groups.size(); }
+
+  // Check that there are still submodules to split.
+  bool hasMoreSplits() const { return remainingSplits() > 0; }
+};
+
+std::unique_ptr<ModuleSplitterBase>
+getDeviceCodeSplitter(ModuleDesc MD, IRSplitMode Mode, bool IROutputOnly,
+                      bool EmitOnlyKernelsAsEntryPoints);
+
+void dumpEntryPoints(raw_ostream &OS, const EntryPointSet &C,
+                     std::string_view Msg = "");
+void dumpEntryPoints(raw_ostream &OS, const Module &M,
+                     bool OnlyKernelsAreEntryPoints = false,
+                     std::string_view Msg = "");
+
+struct SYCLSplitModule {
```

Comment:
Please consider adding a descriptive comment. Thanks

Reply:
We are going to add Properties into this structure in follow-up patches.